### ISSUE TITLE
Restructuring the files on the main repo

### ISSUE THREAD

**Issue by rsanfer** [su2code/SU2#583]
Another important topic from the meeting was the need to restructure the files on the main repository, to reduce their sizes and improve both readability and compilation times.

Of course, any file movement will affect the developers with ongoing branches (see #582) so one suggestion is to restructure one file at a time, maybe once every month or so, to reduce the burden to everybody.

I have a number of files in my radar that are priorities:
- driver_structure.cpp (I already have a branch ongoing on this, as some of you know)
- iteration_structure.cpp
- config_structure.cpp
- geometry_structure.cpp
because of their length and the mixture of class structures. There is also a need to improve the includes throughout the code. 

I would like to collect your thoughts.

**Comment by pcarruscag** [comment]
In terms of file organisation I can think of two strategies, one would be grouping them according to parent class, numerics, solver, etc. this would maybe make navigation easier for people who work on the code daily but it does little more than what the naming convention already does. The other would be to group families of classes and that, I think, would help people who are less familiar with the code, so for example numerics, solvers, variables, etc. used for fluid simulations would be separate from the ones used for other physics.

The main re-compilation trigger is probably the config, every file includes it and not all need it. Not much can be done here because most classes do use the config in some way but we could try to minimise how much the dependency spreads by:
- Reviewing what cpp files actually need to include config_structure.hpp
- Including from the cpp and using a forward declaration on the corresponding hpp

Other sensitive "triggers" are the header files of classes with many children (solver, numerics, and variable) splitting the header files for these (maybe by family for small classes to avoid an explosion of hpp files) would allow working on a child class without re-compiling all others so often. The clients of CSolver would only include the header for the abstract class so those too would be shielded from changes in children classes.

**Comment by economon** [comment]
Nice post @pcarruscag : I think we can gain a lot by first breaking up the files, especially the headers, and then being much better about include statements. In particular, we can look at forward declarations, using more guards, and just minimizing the number of includes (I expect some are redundant).

As far as breaking up the files, I think geometry is a good place to start. I can take a shot at this first soon. 

More ideas out there? Anyone have some bandwidth to experiment a little with the headers?

**Comment by economon** [comment]
Bumping this thread...

Files are now starting to be divided in several PRs, but we should set a clear policy for this. Seems that the consensus is to carry one class per file, move the inlines to the headers, and create subfolders where possible (i.e., have a more flat hierarchy in the src directory based on logical groups such as numerics, geometry, solvers, etc). We do not have a clear naming convention yet, but if it is one class per file, then an option is simply the class name.

Am I capturing the current consensus correctly? Anything I am missing? Naming preferences?

**Comment by talbring** [comment]
Yep these are the things I also got so far. If anyone has better idea, please let us know! :+1:

**Comment by clarkpede** [comment]
One comment I have-- @pcarruscag mentioned that it's somewhat difficult to break apart the "config" [blob](https://sourcemaking.com/antipatterns/the-blob).  I have a proposal that could work to break apart things.  Some config options, like those related to grid movement or timestepping or nondimensionalization, are used throughout a lot of the code.  It will be difficult to make headway there.  But some config settings have a very localized impact in the code. To find examples, just look for common prefixes in the `config_template.cfg`.  There's `DV_*`, `FFD_*`, `UQ_*`, `INC_*`, etc.

We could break apart the config class into a group of classes. Each class would derive from the base config class and share its features.  Any function that needs config settings could then pull in two config objects, one with global settings (like nondimensionalization) and one with problem-specific settings (like incompressible flow settings).  That way, changing the problem-specific settings, such as the incompressible flow options, won't trigger a complete rebuild of the code.  It would only trigger a rebuild of the code that uses the incompressible flow options.

That's my idea.  Aside from that idea, I agree with the changes as proposed by @economon.

**Comment by pcarruscag** [comment]
@economon should we have one file per logical group with all the required includes, so that elsewhere you can still write "#include numerics" instead of having to include all one by one?

@clarkpede that would require passing a "matrix" of CConfigs like we do for numerics and so on right? Which would mean making things virtual and so the inlines would not be inline anymore (I think). I like the idea of breaking up the blob... but.

**Comment by clarkpede** [comment]
@pcarruscag Yes, we would have a group (or matrix, I guess) of CConfigs.  We don't have to follow a rigorous "one-config-per-class" definition; we could make a lot of headway just by splitting things up very roughly.

If I understand, you're thinking we would have to have a base class with lots of virtual functions, and then derived classes with specific implementations.  I don't think that would be necessary.  First off, that wouldn't solve any problems, since the base class would have to have virtual placeholders for all the config classes.  Second, the respective solver classes know what calls they need to make.  So they can just have a specific derived class as a parameter (e.g. `CIncSolverConfig`).  They don't need an abstract base class because we expect the incompressible settings to incompressible settings, and we won't ever switch it out for any other type of settings.  The derived classes (e.g. `CIncSolverConfig`) could implement their own config getters/setters, without having any matching virtual functions in the base Config class.  That should allow us to keep our inlines and avoid dynamic dispatches.

There may be other flaws with my thinking that i'm not seeing.  I just wanted to put my idea out there so we can make more informed decisions.  I can also move this discussion to a separate issue, if we want to flesh out this discussion more.

**Comment by economon** [comment]
@pcarruscag : as for header issues, I believe the consensus was also moving toward completely splitting the headers by class as well. It is true that then we may need to include multiple headers one-by-one, but in most cases, I think that we may just need to add a handful (rather than all), and this division should also help us avoid some of the dependencies during compile time. Open to more opinions on this, of course.

**Comment by pcarruscag** [comment]
What I mean, taking numerics as example, is (I am not sure I explained it well):
> Numerics
>> numerics.hpp
>> CNumerics.hpp
>> CCentJST_Flow.hpp
>> ...
In "numerics.hpp" you put all the includes so that in solver you do #include "numerics.hpp" instead of repeating the list (which in this case is quite long).

**Comment by talbring** [comment]
I hope that we only have to include the specific headers in `driver_structure.cpp` and everywhere else just the `CNumerics.hpp`.

**Comment by economon** [comment]
Same. If we construct it correctly, I am hoping that we will actually only need to include certain individual headers in the solvers, etc. (or maybe just the parent numerics header w virtual function descriptions) so that we can break some of the dependencies that lead to longer build time.

**Comment by WenyinWei** [comment]
Yeah, some class files are so long that even the compilers will complain. I once compiled a cross-platform project in Windows and the VS compiler complained about the length of several files in the project had exceeded the limit(they are about 4000 lines). I have to manually edit the compiler length limit to make it tolerate the codes. 

**driver_structure.cpp** has about 8000 lines in the master branch. Really horrible. Luckily you have restructured them. That must be a nightmare for newcomers. Splitting the class files into pieces composed of parent class and its classes and putting them in a subfolder will be deeply appreciated.

**Comment by WenyinWei** [comment]
One more detail, it seems that all the pointers pointing to new arrays are initialized with an extra loop. I guess that we can use the following way to make the code concise, without an extra loop and no need to specify the loop range manually. 

```
int *pia = new int[10]; // Without initialization 
int *pia2 = new int[10] ();  // Initialized with 0
```

**Comment by talbring** [comment]
@WenyinWei We have not enabled C++11 throughout the whole code yet. We just now have a PR open which enables it by default. So we might consider that way of initializing arrays in the future.

**Comment by WenyinWei** [comment]
Hello @talbring , surprisingly when I was coding and hanging out online casually, I found the initialization feature is not C++11 but C++03. You could have a check the [stackoverflow page](https://stackoverflow.com/questions/2204176/how-to-initialise-memory-with-new-operator-in-c). That's the first answer.

**Comment by talbring** [comment]
@WenyinWei Thanks for the information. This is indeed a lesser known feature ... But its useful and  will should try to apply it more often now.

**Comment by pcarruscag** [comment]
WIP... nearly there but wip...

**Comment by economon** [comment]
We are getting closer on this topic, but I am still wondering if we can go further... 

I think it would be great if we can eventually do away entirely with the first two levels of the directory structure: the Common/, SU2_CFD/, etc. level as well as the src/, obj/, etc. level below that. Instead, we can have all cpp and hpp alongside each other in the various subdirectories at the root level (or maybe inside a single cpp/ or src/ directory at the root level). 

At the moment, a lot of navigation is required to move between files that are logically related due to the module structure and the separate of cpp/hpp files.

**Comment by pcarruscag** [comment]
What do you mean by logically related? [Compressible / Incompressible / Turbulent] solver+variable+numerics+output ?
I think having the top level of folders following the class hierarchy is helpful to someone who is trying to follow the code execution for the first time.
Maybe we can have some subdirs for specific physics in /solvers /variables and /output (similar to what I did with /numerics to keep a compromise between the old file structure and the new)?
I would not mind merging Common and SU2_CFD now that there are descriptive sub-directories for geometry, solvers, etc. I'm not sure if linking everything into just one library is a good idea though (e.g. is it going to make the binaries much larger?).
We should perhaps then move the applications to a separate folder "SU2_APPS" where we would put CFD, DEF, DOT etc.
I don't mind having hpp's and cpp's together as we have a fair bit of implementation also in header files, I guess most IDE's have filters to separate them if folks prefer to see them that way.

**Comment by economon** [comment]
To be clear: I like the solver/, variable/, output/, etc. layout a lot, and I am simply suggesting that we move entirely to that layout (with cpp and hpp merged together in each of those folders) and merging the code in Common, SU2_CFD, and other modules into that structure too (in their own folders still, like geometry/ for example).

To keep it organized you can simply make each of those directories a library in meson which all later get linked into the various binaries. The mains could live in the top level src/ directory if we keep it or a separate directory (open for me). Something like

```
SU2/
  src/
    solver/
      meson.build
      solver_\*.cpp
      solver_\*.hpp
    variable/
      meson.build
      variable_\*.cpp
      variable_\*.hpp
    ...
```

with or without the src/ directory in the root (could go either way).

**Comment by pcarruscag** [comment]
Got it, by navigation you mean from cpp to hpp then?
I agree with you but I would suggest having those extra subdirs in solver and co.
```
SU2/
  solvers/
    meson.build
    CSolver.hpp
    CSolver.cpp
    flow/
      solver_\*.cpp
      solver_\*.hpp
    turbulence/
```
For me what makes navigation a bit cumbersome is scrolling past all the variables to get to the ones of a particular solver for example, the hpp/cpp I don't mind so much their 2 clicks away in the IDE I use.

**Comment by economon** [comment]
Documenting the plan discussed in our developers call this morning here:
* Put header and sources together 
* Remove separation between Common/, SU2_CFD/, etc.
* Each major folder becomes a library (geometry/, solvers/, etc.)
* Keep some subdirectories in major folders (e.g., solvers/flow/)
* Set up include paths in the meson build system to make #include statements clean in the source
* Will prob require removing automake as part of this, which can’t build these libraries in parallel

Still need to finish separating files first. Tasks:
* @talbring  - iteration classes
* @jayantmukho - grid adaptation, ADT structure, grid movement
* @pcarruscag - AD structure, folders for fem_\* 
* @economon - fluid model

Somewhat related:
* Run a test to see how many formatting-related conflicts result from running clang-format on and then diffing two fairly different branches (@pcarruscag)

Please add anything else I am missing.

**Comment by pcarruscag** [comment]
Couple of issues with clang-format, found in the context of #1044, on this kind of style:

```
su2double
foo, /* blabla */
bar; /* blabla */
```
It goes mental sometimes.

Order of includes is modified, which is guaranteed to break compilation.

So running clang-format on the whole code in one code is maybe possibly perhaps definitely not safe.

**Comment by pcarruscag** [comment]
Apart from those issues I have to say I'm enjoying the freedom of just cutting and dumping code from place to place with no care for formatting, more than I expected.

**Comment by economon** [comment]
👍 - care free formatting is a huge plus. I couldn't agree more.

Thanks for the notes on the issues.. we can clean some of those up. At least for the order of includes, I am guessing there is a clang-format option to disable that. I've been changing the comma-comment declarations you mention to all have semi-colons lately when I come across them, since that makes it easier to add or remove individual lines.

**Comment by pcarruscag** [comment]
@ALL we are aiming to complete this work (see Tom's comment on the 20th of May) **within 3 months**.
In so doing we will also **discontinue the legacy build system** (pending some testing on machines without internet connection).

So, if you have branches with large changes to existing files, bring them to develop ASAP to minimize conflicts.
If you have new files, the conflicts will be less severe (just include paths probably).

I will try to script this file movement and subsequent meson updates, so that in the worst case folks can do the same modifications on their branches automatically and then merge the also modified develop.
But, because we also **want to put clangformat in place**, I REALLY encourage you to merge your work into develop soon!
It is the perfect time to give your code to the world and also to give code reviews to your fellow SU2'ers.

I'll come back with updates and reminders.

P.S. 3 months, tick-tock tick-tock :)

**Comment by erangit** [comment]
Hi,  
I'm worried about discontinuing the legacy build system. While on internet-connected machines the  Meson-Ninja build is superior, I still didn't manage to apply it on clusters without such a connection. In my opinion should keep the configure-make-make install approach alive so that installation is possible for everyone, not just very efficient to part of the community 
Best, Eran
